src/serverenvironment.cpp

```diff
@@ -974,6 +983,39 @@ class ABMHandler
 					}
 					// No required neighbor found
 					continue;
+				} else if (aabm.check_required_neighbors || aabm.check_without_neighbors) {
+					v3s16 p1;
+					bool have_required = false;
+					for(p1.X = p0.X-1; p1.X <= p0.X+1; p1.X++)
+					for(p1.Y = p0.Y-1; p1.Y <= p0.Y+1; p1.Y++)
+					for(p1.Z = p0.Z-1; p1.Z <= p0.Z+1; p1.Z++)
+					{
+						if(p1 == p0)
+							continue;
+						content_t c;
+						if (block->isValidPosition(p1)) {
+							// if the neighbor is found on the same map block
+							// get it straight from there
+							const MapNode &n = block->getNodeNoCheck(p1);
+							c = n.getContent();
+						} else {
+							// otherwise consult the map
+							MapNode n = map->getNode(p1 + block->getPosRelative());
+							c = n.getContent();
+						}
+						if (aabm.check_required_neighbors && !have_required) {
+							if (CONTAINS(aabm.required_neighbors, c)) {
+								have_required = true;
+							}
+						}
+						if (CONTAINS(aabm.without_neighbors, c))
+							goto neighbor_invalid;
+					}
+					if (have_required || !aabm.check_required_neighbors)
```

Comment:
have_required cannot be true if check_required_neighbors is false, right? So that does not have to be checked.